### ISSUE TITLE
`clone_double_ref` doesn't help if the result leads to a compiler error

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust-clippy#7686]
See for example this snippet: 

```rust
struct NonCopy;

fn main() {
    let copy: NonCopy = (&NonCopy).clone();
}
```

It doesn't compile because of the type mismatch (expected struct `NonCopy`, found `&NonCopy`) and the lint isn't shown even though the clone of the double ref is what causes the issue. 

In this small example, it's trivial to spot the error, however, in real code, the error may be caused by something less obvious and a lot harder to debug without knowing that the reference was cloned. For example, I've found a similar problem in generic async code where lifetime mismatch error was very confusing.

Is it possible to run this lint even if the compilation failed?

@rustbot label +C-enhancement

**Comment by dswij** [comment]
I'd imagine that in a non-trivial situation, you'd need to have some information about the type to run this lint, which basically means that `clone_double_ref` can't be run before compilation.

Though I can see how the lint can be ran before compilation for trivial errors such as the one above, I'm not sure if Clippy spotting those trivial errors will be useful enough.

**Comment by rodrigorc** [comment]
I think this should be upgraded to a compiler warning, not a lint. References are `Clone` but also `Copy`, if you have a reference and want another you just copy it. I can't imagine any real code where cloning a reference is actually intended (other than cloning a generic that happens to be a reference, of course).
